app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt

```diff
@@ -24,40 +25,49 @@ import com.naman14.timberx.models.Song
 import com.naman14.timberx.repository.AlbumRepository
 import com.naman14.timberx.repository.ArtistRepository
 import com.naman14.timberx.repository.SongsRepository
-import com.naman14.timberx.util.doAsyncPostWithResult
+import com.naman14.timberx.ui.viewmodels.base.ScopedViewModel
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.Dispatchers.Main
+import kotlinx.coroutines.withContext
 
-class SearchViewModel(val context: Context) : ViewModel() {
+class SearchViewModel(val context: Context) : ScopedViewModel(Main) {
 
     private val searchData = SearchData()
     private val _searchLiveData = MutableLiveData<SearchData>()
 
-    val searchLiveData = _searchLiveData
+    val searchLiveData: LiveData<SearchData> = _searchLiveData
 
     fun search(query: String) {
         if (query.length >= 3) {
-            doAsyncPostWithResult(handler = {
-                SongsRepository.searchSongs(context, query, 10)
-            }, postHandler = {
-                if (it!!.isNotEmpty())
-                    searchData.songs = ArrayList(it)
+            launch {
+                val songs = withContext(IO) {
```

Comment:
The fancy thing about coroutines is it's not just threading. Kotlin generates special code under the hood that handles context switching and line-by-line execution. And it uses thread pools which are FAR cheaper than actual threads.